tools/extra/fpgadiag/nlb0.cpp

```diff
@@ -310,14 +314,56 @@ bool nlb0::setup()
     options_.get_value<bool>("suppress-hdr", suppress_header_);
     options_.get_value<bool>("csv", csv_format_);
 
+    options_.get_value<std::string>("id", nlb0_id_);
+    fpga_guid nlb0_id;
+
+    if (uuid_parse(nlb0_id_.c_str(), nlb0_id)<0) {
+        std::cerr << "Invalid NLB0 id:" << nlb0_id_ << std::endl;
+        return false;
+    }
+
+    uint32_t offset = 0;
+    uint64_t dfh = 0;
+    uint8_t* n = (uint8_t *)&nlb0_id;
+    uint64_t nlb0_hi = swap64(uint64_t, n);
+    uint64_t nlb0_lo = swap64(uint64_t, n+8);
+
+    // enumerate dfh if input id arguement 
+    if (options_.find("id")) {
+
+        bool found = false;
+
+        do {
+             uint64_t feature_uuid_lo, feature_uuid_hi;
+             // Read the next feature header
+             dfh = accelerator_->read_csr64(static_cast<uint32_t>(offset));
+             feature_uuid_lo = accelerator_->read_csr64(static_cast<uint32_t>(offset + 8));
+             feature_uuid_hi = accelerator_->read_csr64(static_cast<uint32_t>(offset + 16));
+             if ((nlb0_lo == feature_uuid_lo) && (nlb0_hi == feature_uuid_hi)) {
+                  offset_ = offset;
+                  found = true;
+                  printf("found the NLB offset=0x%x\n", offset);
+                  break;
+             }
+
+            offset += NEXT_DFH_OFFSET(dfh);
+        } while (!DFH_EOL(dfh));
+
+        if (!found) {
+            std::cerr << "Not found NLB:" << nlb0_id_ << std::endl;
+            return false;
+        }
+
+    }
+
     // TODO: Infer pclock from the device id
     // For now, get the pclock frequency from status2 register
     // that frequency (MHz) is encoded in bits [47:32]
-    uint64_t s2 = accelerator_->read_csr64(static_cast<uint32_t>(nlb0_csr::status2));
+    uint64_t s2 = accelerator_->read_csr64(static_cast<uint32_t>(nlb0_csr::status2) + offset_);
```

Comment:
Instead of adding `+offset_` to all reads/writes, can we or should we make this a member template function in nlb0.h?
```C++
// nlb0.h
template<typename T>
uint32_t read_csr32(T address)
{
    return accelerator_->read_csr32(static_cast<uint32_t>(address)+offset_);
}

template<typename T>
uint64_t read_csr64(T address)
{
    return accelerator_->read_csr32(static_cast<uint32_t>(address)+offset_);
}

template<typename T>
void write_csr32(T address, uint32_t value)
{
    accelerator_->write_csr32(static_cast<uint32_t>(address)+offset_, value);
}

template<typename T>
void write_csr64(T address, uint64_t value)
{
    accelerator_->write_csr64(static_cast<uint32_t>(address)+offset_, value);
}

// nlb0.cpp
// ...
uint64_t s2 = read_csr64(nlb0_csr::status2);

//...
write_csr64(nlb0_dsm::basel, reinterpret_cast<uint64_t>(dsm_->io_address()));

```

Reply:
fixed